### ISSUE TITLE
mpv_opengl_cb_report_flip be called inside a DisplayLink callback

### ISSUE THREAD

**Issue by Akemi** [lhc70000/iina#1542]
**System and IINA version:**
- macOS 10.12.6
- IINA 0.0.15

**Expected behavior:**
`mpv_opengl_cb_report_flip()` should be called consistently on every vsync to get the best result. see documentation in the [header](https://github.com/mpv-player/mpv/blob/7ee01860d9c16f2d9d9ea3f5ae5c2a699c7cf1f9/libmpv/opengl_cb.h#L351). ideally it should be within a display link callback.

**Actual behavior:**
when comparing video-sync settings between iina and mpv i noticed that iina is choppier. looked at the code to find a possible reason and found the following issue. [mpv_opengl_cb_report_flip()](https://github.com/lhc70000/iina/blob/fa8d337cb0402c026e8261ea137874b3333475a5/iina/ViewLayer.swift#L160) is called within the draw function of the layer which is called from `display()`. `display()` is usually always called from the [updated callback](https://github.com/lhc70000/iina/blob/fa8d337cb0402c026e8261ea137874b3333475a5/iina/ViewLayer.swift#L22) in iina. because of that `mpv_opengl_cb_report_flip()` isn't actually called on every vsync but rather at every update callback, that is either just the video fps or with `video-sync=display-*` when mpv thinks a vsync happens. in the latter case that won't help mpv to determine a proper vsync since it uses its own data. furthermore the flip is reported after drawing which will introduce an unnecessary delay and jitter for example in the case the renderer needs to change settings. additionally there is also pause or temporary bad performance that could delay or stop the callback. also keep in mind the layer refuses to draw when being offscreen (minimised, different space, etc) and the [draw](https://github.com/lhc70000/iina/blob/fa8d337cb0402c026e8261ea137874b3333475a5/iina/ViewLayer.swift#L129) function is not being called after `display()`, which in itself also causes problems like deactivating `display-resample`.

related to a closed issue #28. this was probably reintroduced when the change to the CAOpenGLLayer was done.

**Steps to reproduce:**
the choppy playback can be reproduced with `video-sync=display-resample`, though i actually don't know if it is caused by the issue described above, or if there is another reason.
- [x] MPV does not have this problem.

**Comment by inflation** [comment]
I found the bug long time ago but don't know what causes it. I opened an issue at #741, and further I found that if you don't specify `display-fps`, it would report 0. However if you do, it results in my screenshot in the issue.

Thanks for the head up, would try to see if anything improves.

**Comment by Akemi** [comment]
yeah i kept an eye on that issue some time ago too. i might be able to clarify one issue, `display-fps` being 0 when not set. this is expected because retrieval of the display fps is not part of the opengl-cb API and would need to be implemented in iina itself by setting the `display-fps` accordingly. a quick search on the iina rep seems like it doesn't do this(?).

but i see now, display resample was never really active because the display fps was not set, no wonder it was choppy. though i can also reproduce your issue with drastically too high and increasing estimated display-fps when activated.

i actually didn't check this because i had problems getting `stats.lua` to work with iina.

**Comment by lhc70000** [comment]
Thanks a lot, I'll have a look soon.

**Comment by Akemi** [comment]
i just tested the proposed change with iina and have to take back what i said on issue #741. this [patch](https://0x0.st/saox.diff) (the proposed fix) will fix issue #741 and #125, when setting `display-fps` manually to a somewhat accurate value. the patch itself was just a test and obviously needs a proper implementation, proper start/stop, retrieval of the actual display fps, etc.

anyway this was bugging me the whole night and i wanted to know the reason.

**Comment by lhc70000** [comment]
Sorry for the delay. I've added this in `mpv-new-api` branch several days ago together with implementation of mpv's new render API. Seems that it solves #741 according to @inflation 's test.

**Comment by Akemi** [comment]
sry i can't test it since i am still on 10.12.6 which doesn't support xcode 9.3 and newer swift versions. code wise it looks okay, thought two remarks.
1. you should probably set `display-fps` to the current refresh rate on init and when the display setting change. so mpv knows where to start with estimating. otherwise it will unnecessary drop frames.
2. display-resample will 'break' (deactivates itself because the render function isn't called anymore) when iina is off screen (different space, minimised) since the layer will refuse to draw. with the advanced rendering control activated you can manually call the render function with the [skip](https://github.com/Akemi/mpv/commit/c5b483692678910d33de5e8ee1b589c85e8ba7de#diff-d04b222e1aa47981e63ea56f2be5519eR134) parameter set. the condition for that is a `display()` call without a corresponding `draw(inCGLContext...) `call. this is also a new feature of the new rendering API.
